python/lsst/pipe/drivers/skyCorrection.py

```diff
@@ -0,0 +1,368 @@
+            Data identifier, with `dataId` (data identifier) and `bgModel`
+            (background model) elements.
+
+        Returns
+        -------
+        bgModel : `lsst.pipe.drivers.background.FocalPlaneBackground`
+            Background model.
+        """
+        assert cache.dataId == data.dataId
+        data.bgModel.addCcd(cache.exposure)
+        return data.bgModel
+
+    def subtractModel(self, cache, dataId, bgModel):
+        """Subtract background model
+
+        This method runs on the slave nodes.
+
+        Parameters
+        ----------
+        cache : `lsst.pipe.base.Struct`
+            Process pool cache.
+        dataId : `dict`
+            Data identifier.
+        bgModel : `lsst.pipe.drivers.background.FocalPlaneBackround`
+            Background model.
+
+        Returns
+        -------
+        exposure : `lsst.afw.image.Exposure`
+            Resultant exposure.
+        """
+        assert cache.dataId == dataId
+        exposure = cache.exposure
+        image = exposure.getMaskedImage()
+        detector = exposure.getDetector()
+        bbox = image.getBBox()
+        cache.bgModel = bgModel.toCcdBackground(detector, bbox)
+        image -= cache.bgModel.getImage()
+        cache.bgList.append(cache.bgModel[0])
+        return self.collect(cache)
+
+    def collect(self, cache):
+        """Collect exposure for potential visualisation
+
+        This method runs on the slave nodes.
+
+        Parameters
+        ----------
+        cache : `lsst.pipe.base.Struct`
+            Process pool cache.
+
+        Returns
+        -------
+        detId : `int`
+            Detector identifier.
+        image : `lsst.afw.image.MaskedImage`
+            Binned image.
+        """
+        return (cache.exposure.getDetector().getId(),
+                afwMath.binImage(cache.exposure.getMaskedImage(), BINNING))
+
+    def collectOriginal(self, cache, dataId):
+        """Collect original image for visualisation
+
+        This method runs on the slave nodes.
+
+        Parameters
+        ----------
+        cache : `lsst.pipe.base.Struct`
+            Process pool cache.
+        dataId : `dict`
+            Data identifier.
+
+        Returns
+        -------
+        detId : `int`
+            Detector identifier.
+        image : `lsst.afw.image.MaskedImage`
+            Binned image.
+        """
+        exposure = cache.butler.get("calexp", dataId, immediate=True)
+        return (exposure.getDetector().getId(),
+                afwMath.binImage(exposure.getMaskedImage(), BINNING))
+
+    def write(self, cache, dataId):
+        """Write resultant exposure
+
+        This method runs on the slave nodes.
+
+        WARNING: We clobber the calexp in the data repository! This may not
+        be desirable, but nor do we want to introduce multiple datasets that
+        the user has to select down the road.  The user should write to a
+        different rerun or output data repository.
```

Comment:
It's true the user can write to a different output repo, but I would rather introducing another butler dataset type, and let the user select which one to use later.  Overwriting data files make it very difficult for provenance, workflow, to name a few.  Maybe butler+supertask will come up with some brilliant solutions later, but before that happens I do not like the idea of sharing dataset type names.